### ISSUE TITLE
[Snaps] Filesystem errors when loading a snapshot on a different machine

### ISSUE THREAD

**Issue by CuriousGeorgiy** [firecracker-microvm/firecracker#4036]
# Describe the bug

I am hacking on firecracker-containerd to support firecracker snapshots, and I am facing the following problem.

I create a snapshot of a microVM running a nginx container, and then I try to load this snapshot on a different machine (for preserving the disk state, I simply commit a container snapshot).

When restoring the disk state, I [patch](https://github.com/firecracker-microvm/firecracker/issues/4014#issuecomment-1677099508) (see also #4014) the snapshot's disk device path for the container snapshot device to point to a fresh containerd snapshot mount point (the previously committed image is mounted).

Snapshot loading succeeds and the container is responsive even to http requests (I am dropping the network setup details since I don't have problems with it now), but the nginx container returns internal server errors, and the following error appears in the microVM's kernel logs:

```console
DEBU[2023-08-15T07:21:33.072516371-04:00] [   38.858355] EXT4-fs error (device vdb): ext4_find_entry:1447: inode #262: comm nginx: checksumming directory block 0  jailer=noop runtime=aws.firecracker vmID=1 vmm_stream=stdout
DEBU[2023-08-15T07:21:33.076584363-04:00] [   38.862540] EXT4-fs error (device vdb): ext4_find_entry:1447: inode #262: comm nginx: checksumming directory block 0  jailer=noop runtime=aws.firecracker vmID=1 vmm_stream=stdout
```

It seems like the restored disk state (via container snapshot commit) is inconsistent with the microVM's disk state.

In this regard, do you have any ideas of what the problem is (if it doesn't simply come down to state disk state inconsistency) and how it could be solved? It seems like this requires a sort of a filesystem checkpoint, but I couldn't find any relevant information on the web.

This issue is rather a question than a bug report, so I am not adding any other context (though I can if it will help clarify my use case).

## Checks

- [x] Have you searched the Firecracker Issues database for similar problems?
- [x] Have you read the existing relevant Firecracker documentation?
- [x] Are you certain the bug being reported is a Firecracker issue?

**Comment by roypat** [comment]
Hi, 
thanks for opening this issue! Firecracker does not support renaming resources at snapshot-restore. While we are aware of this limitation, we also have no explored yet what changes to firecracker would be required to properly support this.
Can you reproduce issue using an unpatched firecracker binary? E.g. trying to restore a snapshot where the path to the disk's backing file on the new machine is identical to the path on the machine on which the snapshot was taken.

**Comment by CuriousGeorgiy** [comment]
@roypat 

Hi! Okay, I will try tinkering around with mounting to achieve this, though I don't catch how the block device path can affect functionality.

Also, some additional context just in case, I save the disk state the following way (step 4 should actually be redundant, but I added it just in case):
1. Pause microVM
2. Create microVM snapshot
3. `dmsetup suspend` on the corresponding container mount
4. `ioctl BLKFLSBUF` on the corresponding container mount
5. `dmsetup resume` on the corresponding container mount
6. an analog of `docker commit` for containerd (it simply creates a new container image that accounts for the RW (topmost) layer of the container and pushes it to a local registry)
7. Resume microVM

I can provide a code snippet for clarity if needed.

**Comment by bchalios** [comment]
Hey @CuriousGeorgiy some thoughts:

Is there any chance that these:

3. `dmsetup suspend` on the corresponding container mount
4. `ioctl BLKFLSBUF` on the corresponding container mount

modify the contents of the file backing the drive?

Also, could you try to perform a flush operation in the guest and ensure it completes before pausing the microVM and taking the snapshot?

**Comment by CuriousGeorgiy** [comment]
@bchalios 

No, they only flush the disk state. 

As per the `dmsetup suspend` man page [^1]:

>  Suspends a device.  Any I/O that has already been mapped
              by the device but has not yet completed will be flushed.
              Any further I/O to that device will be postponed for as
              long as the device is suspended.

As per `ioctl BLKFLSBUF`, it is supposed to flush the buffer cache for the device[^2].

I also added an explicit `fsync` on the device, but it also didn't help.

> Also, could you try to perform a flush operation in the guest and ensure it completes before pausing the microVM and taking the snapshot?

AFAIC, this is not enough, since disk state is apparently changed during snapshot creation: if I suspend the container snapshot device (which blocks I/O) and take a snapshot of the microVM, the firecracker process hangs in `D` state. 

So, what I am doing right now, is trying to flush the guest _after_ creating a snapshot of the microVM _while_ it is still in _paused_ state.

[^1]: https://man7.org/linux/man-pages/man8/dmsetup.8.html
[^2]: https://github.com/torvalds/linux/blob/4853c74bd7ab7fdb83f319bd9ace8a08c031e9b6/tools/include/uapi/linux/fs.h#L152

**Comment by CuriousGeorgiy** [comment]
> Also, could you try to perform a flush operation in the guest and ensure it completes before pausing the microVM and taking the snapshot?

Sorry, looks like I got your question wrong. What do mean by performing a flush operation in the guest? If you mean manually flushing files, than:
1. How is this generalizable to any container?
2. I tried running a simple Golang http server (no filesystem manipulation), and when I restore the container, and send a request, it crashes (panics), so I think the problem is not with file flushing.

**Comment by bchalios** [comment]
> How is this generalizable to any container?

It is not, on files but you can do that for the drives using `sync`. In any case, I was just trying to root-cause. My thinking went: maybe while pausing the microVM some write operation was on-the-fly, and maybe this triggers a corner-case we don't handle, i.e. the write was interrupted and the guest does not know how to handle this.

Generally speaking, we mainly test the snapshotting feature with guests in a sort of "quiesced" state, so there might be a corner case we are missing.

**Comment by CuriousGeorgiy** [comment]
@bchalios 

> It is not on files, but you can do that for the drives using `sync`.

Yeah, that's what I am already trying to achieve using `dmsetup suspend`, `fsync` and `ioctl BLKFLSBUF` (see [comment above](https://github.com/firecracker-microvm/firecracker/issues/4036#issuecomment-1682322577)).


> My thinking went: maybe while pausing the microVM some write operation was on-the-fly, and maybe this triggers a corner-case we don't handle, i.e. the write was interrupted and the guest does not know how to handle this.

I expect this to be handled by the snapshot creation request (and, as I noticed [above](https://github.com/firecracker-microvm/firecracker/issues/4036#issuecomment-1682322577), I observe some I/O on the container mount drive during snapshot creation). 

Is there anything I could do to try to narrow down the problem and make a simple reproducer? As I [said](https://github.com/firecracker-microvm/firecracker/issues/4036#issuecomment-1683526453), I tried running a very simple Golang http server and try do some filesystem I/O, but it just crashes even without filesystem manipulation.

**Comment by bchalios** [comment]
Could you try to reproduce the same scenario without `firecracker-containerd`?:

1. Launch a firecracker microVM "by hand", you could follow the [Getting Started Guide](https://github.com/firecracker-microvm/firecracker/blob/main/docs/getting-started.md)
2. Launch your Go server inside the microVM.
3. Pause & snapshot
4. Resume from the snapshot

And see if it reproduces. After step 4. you could also try to patch the drive to another path, so that we see if that's the problem.

**Comment by bchalios** [comment]
Regarding the last comment, you can:

4a. LoadSnapshot (without resume)
4b. Patch the drive file path on the host
4c. Resume the microVM.

**Comment by CuriousGeorgiy** [comment]
I have tried what you suggested with a RW rootfs, and didn't encounter any problems, everything works okay. 

> Regarding the last comment, you can:
>
>4a. LoadSnapshot (without resume)
4b. Patch the drive file path on the host
4c. Resume the microVM.

I haven't yet tried doing the same stuff on an attached block device like firecracker-containerd does.

In the meanwhile, I have tried adding patching the firecracker-go-sdk and adding patch drive requests in between loading a snapshot without resuming and resuming the microVM, and got the following errors in the logs:
```console
DEBU[2023-08-20T15:15:59.964532664-04:00] Running handler fcinit.LoadSnapshot           runtime=aws.firecracker
INFO[2023-08-20T15:15:59.973124142-04:00] Updating drive /var/lib/firecracker-containerd/runtime/default-rootfs.img, slot root_drive, root true.  runtime=aws.firecracker
INFO[2023-08-20T15:15:59.973851533-04:00] Updated drive /var/lib/firecracker-containerd/runtime/default-rootfs.img: [PATCH /drives/{drive_id}][204] patchGuestDriveByIdNoContent   runtime=aws.firecracker
DEBU[2023-08-20T15:15:59.973889025-04:00] updateDrive returned for /var/lib/firecracker-containerd/runtime/default-rootfs.img  runtime=aws.firecracker
INFO[2023-08-20T15:15:59.973939266-04:00] Updating drive /var/lib/firecracker-containerd/shim-base/vanilla-firecracker-snapshots#0/ctrstub0, slot MN2HE43UOVRDA, root false.  runtime=aws.firecracker
INFO[2023-08-20T15:15:59.974400352-04:00] Updated drive /var/lib/firecracker-containerd/shim-base/vanilla-firecracker-snapshots#0/ctrstub0: [PATCH /drives/{drive_id}][204] patchGuestDriveByIdNoContent   runtime=aws.firecracker
DEBU[2023-08-20T15:15:59.974444103-04:00] updateDrive returned for /var/lib/firecracker-containerd/shim-base/vanilla-firecracker-snapshots#0/ctrstub0  runtime=aws.firecracker
DEBU[2023-08-20T15:15:59.975089233-04:00] VM resumed successfully                       runtime=aws.firecracker
DEBU[2023-08-20T15:15:59.975147605-04:00] snapshot loaded successfully                  runtime=aws.firecracker
INFO[2023-08-20T15:15:59.975205975-04:00] calling agent                                 runtime=aws.firecracker vmID=0
DEBU[2023-08-20T15:15:59.978783082-04:00] [    4.331801] virtio_blk virtio0: new size: 125800 512-byte logical blocks (64.4 MB/61.4 MiB)  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:15:59.981208538-04:00] [    4.334353] virtio_blk virtio1: new size: 1 512-byte logical blocks (512 B/512 B)  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:15:59.982562932-04:00] [    4.335939] vdb: detected capacity change from 10737418240 to 512  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:15:59.994733454-04:00] [    4.348439] VFS: busy inodes on changed media or resized disk vdb  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
INFO[2023-08-20T15:16:00.077125413-04:00] successfully started the VM                   runtime=aws.firecracker vmID=0
DEBU[2023-08-20T15:16:00.077233305-04:00] event published                               ns=vanilla-firecracker-snapshots runtime=aws.firecracker topic=/firecracker-vm/start type=VMStart
DEBU[2023-08-20T15:16:00.077696072-04:00] event forwarded                               ns=vanilla-firecracker-snapshots topic=/firecracker-vm/start type=VMStart
DEBU[2023-08-20T15:16:02.527490728-04:00] [    6.880446] attempt to access beyond end of device  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:16:02.529025777-04:00] [    6.882101] vdb: rw=2049, want=8650760, limit=1  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:16:02.531628879-04:00] [    6.883647] Buffer I/O error on dev vdb, logical block 1081344, lost sync page write  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:16:02.534177637-04:00] [    6.886330] JBD2: Error -5 detected when updating journal superblock for vdb-8.  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:16:02.535708470-04:00] [    6.888824] Aborting journal on device vdb-8.  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:16:02.537336420-04:00] [    6.890319] attempt to access beyond end of device  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:16:02.538357728-04:00] [    6.891948] vdb: rw=133121, want=8650760, limit=1  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:16:02.540002989-04:00] [    6.892971] Buffer I/O error on dev vdb, logical block 1081344, lost sync page write  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
DEBU[2023-08-20T15:16:02.541503916-04:00] [    6.894613] JBD2: Error -5 detected when updating journal superblock for vdb-8.  jailer=noop runtime=aws.firecracker vmID=0 vmm_stream=stdout
```

Can the whole problem be related to data integrity issues with attachable drives that are described [here](https://github.com/firecracker-microvm/firecracker/blob/main/docs/api_requests/patch-block.md#in-flight-io-requests)?

**Comment by bchalios** [comment]
It could be, yes. Are these drives mounted when you resume the microVM? I guess, so right? If that's the case, is there a way we could unmount them before snapshotting?

**Comment by CuriousGeorgiy** [comment]
@bchalios 

Everything also works when mounting a regular file (formatted as a filesystem) as a drive and running the server there. AFAIC, the only different thing firecracker-containerd does is it uses devmapper for creating drives, I will try this out next. 

> Are these drives mounted when you resume the microVM? I guess, so right? If that's the case, is there a way we could unmount them before snapshotting?

AFAIC, yes. I guess I could try hacking into containerd and manually unmounting these drives as a PoC, but I'm not sure this won't break anything inside containerd.

**Comment by CuriousGeorgiy** [comment]
@bchalios 

> AFAIC, to a rough approximation, the only thing that firecracker-containerd does differently (and the last thing I can do to try to narrow down this issue to firecracker) is it uses devmapper for creating drives (i.e., it uses virtual block devices), I will try setting this up tomorrow.

I also tried using a thin pool device for the drive, like firecracker-containerd does, and everything also works fine. 

I guess further investigation requires digging into firecracker-containerd internals, and we should move this issue there. What do you think?

**Comment by bchalios** [comment]
@CuriousGeorgiy thanks a lot for the effort put in this. I think it makes sense, yes. It looks like it's not a Firecracker problem, so the firecracker-containerd folks might be better suited to help you with this.

Feel free to move it. You can always open a new issue here if you find something that is Firecracker related.

**Comment by bchalios** [comment]
I 'll close this now since the issue is tracked in https://github.com/firecracker-microvm/firecracker-containerd/issues/759.

Feel free to re-open if you realize the problem is with Firecracker after all.
